File: py800mon/monitor/appstate.py
import dataclasses
import enum

from ..datastructures import CpuState, DisplayList


class AppMode(enum.Enum):
    NORMAL = 1
    DEBUG = 2
    SHUTDOWN = 3


@dataclasses.dataclass
class AppStateData:
    dlist: DisplayList
    cpu: CpuState
    cpu_disasm: str
    monitor_frame_time_ms: int
    paused: bool
    emu_ms: int
    reset_ms: int
    crashed: bool
    state_seq: int
    last_rpc_error: str | None
    active_mode: AppMode
    ui_frozen: bool
    disassembly_enabled: bool
    disassembly_addr: int | None
    dmactl: int
    breakpoints_supported: bool


_state = AppStateData(
    dlist=DisplayList(),
    cpu=CpuState(),
    cpu_disasm="",
    monitor_frame_time_ms=0,
    paused=False,
    emu_ms=0,
    reset_ms=0,
    crashed=False,
    state_seq=0,
    last_rpc_error=None,
    active_mode=AppMode.NORMAL,
    ui_frozen=False,
    disassembly_enabled=True,
    disassembly_addr=None,
    dmactl=0,
    breakpoints_supported=False,
)


class StateStore:
    def __init__(self, backing: AppStateData):
        self._s = backing

    def set_active_mode(self, mode: AppMode):
        self._s.active_mode = mode

    def set_ui_frozen(self, enabled: bool):
        self._s.ui_frozen = enabled

    def set_disassembly_enabled(self, enabled: bool):
        self._s.disassembly_enabled = enabled

    def set_disassembly_addr(self, addr: int):
        self._s.disassembly_addr = addr & 0xFFFF

    def set_breakpoints_supported(self, enabled: bool):
        self._s.breakpoints_supported = enabled

    def set_status(
        self, paused: bool, emu_ms: int, reset_ms: int, crashed: bool, state_seq: int
    ):
        self._s.paused = paused
        self._s.emu_ms = emu_ms
        self._s.reset_ms = reset_ms
        self._s.crashed = crashed
        self._s.state_seq = state_seq

    def set_last_rpc_error(self, error: str | None):
        self._s.last_rpc_error = error

    def set_cpu(self, cpu: CpuState):
        self._s.cpu = cpu

    def set_cpu_disasm(self, cpu_disasm: str):
        self._s.cpu_disasm = cpu_disasm

    def set_dlist(self, dlist: DisplayList, dmactl: int):
        self._s.dlist = dlist
        self._s.dmactl = dmactl

    def set_frame_time_ms(self, ms: int):
        self._s.monitor_frame_time_ms = ms


_store = StateStore(_state)
store = _store


class StateProxy:
    def __init__(self, backing: AppStateData):
        self._backing = backing

    def __getattr__(self, name):
        return getattr(self._backing, name)

    def __setattr__(self, name, value):
        if name == "_backing":
            object.__setattr__(self, name, value)
            return
        raise AttributeError(
            "State is read-only. Use ActionDispatcher to update state."
        )


state = StateProxy(_state)

File: go800mon/a800mon/monitor/appstate.go
package monitor

import (
	"sync"

	. "go800mon/a800mon"
	"go800mon/internal/displaylist"
)

type AppMode int

const (
	AppModeNormal AppMode = iota + 1
	AppModeDebug
	AppModeShutdown
)

type AppStateData struct {
	DList                displaylist.DisplayList
	CPU                  CPUState
	CPUDisasm            string
	MonitorFrameTimeMS   int
	Paused               bool
	EmuMS                uint64
	ResetMS              uint64
	Crashed              bool
	StateSeq             uint64
	LastRPCError         string
	ActiveMode           AppMode
	UIFrozen             bool
	UseATASCII           bool
	DisassemblyEnabled   bool
	DisassemblyAddr      *uint16
	DMACTL               byte
	History              []CpuHistoryEntry
	DisassemblyRows      []DisasmRow
	BreakpointsSupported bool
}

type DisasmRow struct {
	Addr           uint16
	Size           int
	RawText        string
	AsmText        string
	Mnemonic       string
	Operand        string
	Comment        string
	OperandAddrPos [2]int
	HasOperandAddr bool
	FlowTarget     *uint16
}

type ScreenRow struct {
	Addr uint16
	Data []byte
}

type WatcherRow struct {
	Addr      uint16
	Value     byte
	NextValue byte
	Comment   string
}

type BreakpointConditionRow struct {
	CondType byte
	Op       byte
	Addr     uint16
	Value    uint16
}

type BreakpointClauseRow struct {
	Conditions []BreakpointConditionRow
}

type StateStore struct {
	mu sync.RWMutex
	s  AppStateData
}

var store = newStateStore()

func newStateStore() *StateStore {
	return &StateStore{
		s: AppStateData{
			ActiveMode:         AppModeNormal,
			UseATASCII:         true,
			DisassemblyEnabled: true,
		},
	}
}

func State() AppStateData {
	store.mu.RLock()
	defer store.mu.RUnlock()
	return store.snapshotLocked()
}

func (s *StateStore) snapshotLocked() AppStateData {
	st := s.s
	if st.History != nil {
		h := make([]CpuHistoryEntry, len(st.History))
		copy(h, st.History)
		st.History = h
	}
	if st.DisassemblyRows != nil {
		d := make([]DisasmRow, len(st.DisassemblyRows))
		copy(d, st.DisassemblyRows)
		st.DisassemblyRows = d
	}
	return st
}

func (s *StateStore) setStatus(paused bool, emuMS, resetMS uint64, crashed bool, stateSeq uint64) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.s.Paused = paused
	s.s.EmuMS = emuMS
	s.s.ResetMS = resetMS
	s.s.Crashed = crashed
	s.s.StateSeq = stateSeq
}

func (s *StateStore) setLastRPCError(text string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.s.LastRPCError = text
}

func (s *StateStore) setCPU(cpu CPUState, cpuDisasm string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.s.CPU = cpu
	s.s.CPUDisasm = cpuDisasm
}

func (s *StateStore) setDList(dl displaylist.DisplayList, dmactl byte) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.s.DList = dl
	s.s.DMACTL = dmactl
}

func (s *StateStore) setHistory(rows []CpuHistoryEntry) {
	s.mu.Lock()
	defer s.mu.Unlock()
	out := make([]CpuHistoryEntry, len(rows))
	copy(out, rows)
	s.s.History = out
}

func (s *StateStore) setDisassemblyRows(rows []DisasmRow) {
	s.mu.Lock()
	defer s.mu.Unlock()
	out := make([]DisasmRow, len(rows))
	copy(out, rows)
	s.s.DisassemblyRows = out
}

func (s *StateStore) setFrameTimeMS(ms int) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.s.MonitorFrameTimeMS = ms
}

func (s *StateStore) setUseATASCII(enabled bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.s.UseATASCII = enabled
}

func (s *StateStore) setDisassemblyEnabled(enabled bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.s.DisassemblyEnabled = enabled
}

func (s *StateStore) setDisassemblyAddr(addr *uint16) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.s.DisassemblyAddr = addr
}

func (s *StateStore) setActiveMode(mode AppMode) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.s.ActiveMode = mode
}

func (s *StateStore) setUIFrozen(enabled bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.s.UIFrozen = enabled
}

func (s *StateStore) setBreakpointsSupported(enabled bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.s.BreakpointsSupported = enabled
}
